### PR TITLE
Java Dataflow library: Add local path-sensitivity

### PR DESCRIPTION
**Pull request by Cornelius-Riemenschneider**
Dataflow paths in functions like 
```
      public void f1(boolean b) {
		Object i = null;
		if (b) {
			i = new Integer(6);
		}
		if (!b) {
			sink(i);
		}
	}
```
are now pruned if they cannot occur.

### CODE REVIEW

**Comment by Cornelius-Riemenschneider** [comment]
@aschackmull @hvitved

File: java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -937,6 +937,52 @@ private predicate localFlowStepPlus(
   )
 }
 
+predicate relevantProperty(Node node1, Configuration config, Property property) {
```
Comment (hvitved) [inline java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll]: Make `private`, and add ql doc, e.g.
```
Holds if `property` is a property that holds at a (reflexive, transitive) local successor node of `node1`.
```

File: java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -324,3 +326,42 @@ predicate isUnreachableInCall(Node n, DataFlowCall call) {
...
+ * that implies that all DataFlowBasicBlock `bb2` where the property holds with
+ * `b.booleanNot()` can *not* be part of that dataflow path.
+ */
+abstract class Property extends TProperty {
+  abstract predicate holdsAt(DataFlowBasicBlock bb, boolean b);
+
+  abstract string toString();
+}
+
+class BooleanSsaVarProperty extends Property, TBooleanSsaVar {
```
Comment (hvitved) [inline java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll]: I would make this `private`.

File: java/ql/test/library-tests/dataflow/path-sensitivity/A.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+			y = x;
+		}
+		if (!b2) {
+			sink(y);
+		}
+	}
+
+	// false positive, where we would need to look at the actual CFG (not just the
+	// dataflow graph) to learn about the property
+	public void f13(boolean b) {
```
Comment (hvitved) [inline java/ql/test/library-tests/dataflow/path-sensitivity/A.java]: This expected false positive is actually not in the test output, do you know why?

Comment (Cornelius-Riemenschneider) [reply java/ql/test/library-tests/dataflow/path-sensitivity/A.java]: The dataflow is already pruned (correctly) without my patch, but I don't know why.

File: java/ql/src/semmle/code/java/dataflow/internal/DataFlowUtil.qll [hunk 4]
```diff
@@ -84,6 +84,17 @@ class Node extends TNode {
...
+  /** Gets the callable in which this node occurs. */
+  DataFlowBasicBlock getBasicBlock() {
+    // TODO check all cases!
+    result = this.asExpr().getBasicBlock() or
+    //result = this.asParameter().getAnAccess().getBasicBlock() or
+    //result = this.(ImplicitVarargsArray).getCall() or
+    result = this.(ImplicitInstanceAccess).getInstanceAccess().getCfgNode().getBasicBlock() or
+    result = this.(MallocNode).getClassInstanceExpr().getBasicBlock()
+    //result = this.(ImplicitPostUpdateNode).getBasicBlock()
+  }
```
Comment (hvitved) [inline java/ql/src/semmle/code/java/dataflow/internal/DataFlowUtil.qll]: Come to think of it, I think this should be moved out as a top-level predicate in `DataFlowPrivate.qll` instead, in order not to pollute the public API. Also, the QL doc needs to be updated.

**Comment by Cornelius-Riemenschneider** [comment]
I updated the pull request.
The code has been rebased again, so all the test cases are in the first commit and the code (including the remarks by Tom) are in the second commit.
The exclusion of guards from the path-sensitive tracking is now more precise than before (see DataFlowPropertyImpl in DataFlowPrivate.qll for details).

@hvitved Can you do the sync+stub implementation after you reviewed the changes?

**Comment by Cornelius-Riemenschneider** [comment]
Anders is right, and the approach from this pull request cannot be generalized.
The testcases derived from this process can be found in #2470 for further use.

**Pull request closed without merging**